src/main/java/com/conveyal/analysis/controllers/OpportunityDatasetController.java

```diff
@@ -140,29 +141,31 @@ private OpportunityDatasetUploadStatus downloadLODES(Request req, Response res)
         final String regionId = req.params("regionId");
         final int zoom = parseZoom(req.queryParams("zoom"));
 
-        // default
-        final String accessGroup = req.attribute("accessGroup");
-        final String email = req.attribute("email");
-        final Region region = Persistence.regions.findByIdIfPermitted(regionId, accessGroup);
+        UserPermissions userPermissions = req.attribute(USER_PERMISSIONS_ATTRIBUTE);
+        final Region region = Persistence.regions.findByIdIfPermitted(regionId, userPermissions.accessGroup);
         // Common UUID for all LODES datasets created in this download (e.g. so they can be grouped together and
         // deleted as a batch using deleteSourceSet)
-        final String downloadBatchId = new ObjectId().toString();
         // The bucket name contains the specific lodes data set and year so works as an appropriate name
         final OpportunityDatasetUploadStatus status = new OpportunityDatasetUploadStatus(regionId, extractor.sourceName);
         addStatusAndRemoveOldStatuses(status);
 
-        taskScheduler.enqueueHeavyTask(() -> {
-            try {
-                status.message = "Extracting census data for region";
-                List<Grid> grids = extractor.censusDataForBounds(region.bounds, zoom);
-                createDatasetsFromPointSets(
-                        email, accessGroup, extractor.sourceName, downloadBatchId, regionId, status, grids
-                );
-            } catch (IOException e) {
-                status.completeWithError(e);
-                LOG.error("Exception processing LODES data: " + ExceptionUtils.stackTraceString(e));
-            }
-        });
+        SpatialResource source = SpatialResource.create(userPermissions, extractor.sourceName)
+                .withRegion(regionId);
+
+        taskScheduler.enqueue(Task.create("Extracting LODES data")
+                .forUser(userPermissions)
+                .setHeavy(true)
+                .withWorkProduct(source)
```

Comment:
As above when processing aggregation areas, the source dataset is not the work product -- we're potentially producing many work products that are not known in advance, a situation not covered by the current taskScheduler programming API. We'll need to come up with a way to cover this situation.

Reply:
We decided in today's call that we will introduce an intermediate "group of uploaded file derivatives" table in the database. Processing one uploaded file into one set of derivatives corresponds to exactly one work product in this new table.